### PR TITLE
Update and expand the section on distributed hypertables

### PR DESCRIPTION
**Pull request by erimatnor**
This change updates the section on distributed hypertables with more
information about inserting and querying data, managing data nodes,
and native replication.

### CODE REVIEW

File: using-timescaledb/distributed-hypertables.md [hunk 1]
```diff
@@ -1,41 +1,47 @@
...
+joins between a local table and a distributed hypertable requires
+fetching the raw data from data nodes and performing the join locally.
+
+>:WARNING: Distributed hypertables are not yet recommended for
+production use. For more information, please review the current
+[limitations][distributed-hypertable-limitations]. You can also
+[contact us][contact] or join the #multinode channel in our [community
+Slack][slack].
+
+### Creating a Distributed Hypertable [](create)
```
Comment (mfreed) [inline using-timescaledb/distributed-hypertables.md:39]: Should this be ## or ###, @solugebefola ?  I think these should all just be one level below the top, so ### -> ##, and #### -> ###?  (But I remember with API there was something funny, so letting solar opine).

**Review by mfreed — APPROVED** [review]
Some suggestions, but overall it's a really great, clear overview of distributed hypertables!

File: using-timescaledb/distributed-hypertables.md [hunk 2]
```diff
@@ -48,32 +54,175 @@ CREATE TABLE conditions (
...
+frequently in the `GROUP BY` clause of queries, then it is likely a
+good candidate column for space partitioning.
+
+With space partitioning, multi-row inserts can often be split across
+the data nodes for increased insert performance, and queries can be
+parallelized by executing a subset of the query on each data
+node. TimescaleDB contains many optimizations to take advantage of the
+partitioning scheme to make best use of multi-node resources.
+
+While a distributed hypertable works with only time partitioning it
```
Comment (solugebefola) [inline using-timescaledb/distributed-hypertables.md:92]:
```suggestion
When a distributed hypertable works with only time partitioning it
```

File: using-timescaledb/distributed-hypertables.md [hunk 3]
```diff
@@ -48,32 +54,175 @@ CREATE TABLE conditions (
...
+
+### Inserting into a Distributed Hypertable [](insert)
+
+Just like a regular hypertable, it is important to batch data when
+inserting into a distributed hypertable. Each insert statement is its
+own transaction, and the overhead and cost of this transaction is good
+to amortize over many rows of data. With a distributed hypertable, the
+transaction has additional costs due to the coordination that needs to
+happen across data nodes (e.g., two-phase commit protocol). There are
+two ways to insert data:
```
[the remark this answers is not in the document]
Comment (solugebefola) [reply using-timescaledb/distributed-hypertables.md:107]: @mfreed the two parentheticals don't seem necessary.  The second parenthetical uses 'similar' twice in close proximity

File: using-timescaledb/distributed-hypertables.md [hunk 4]
```diff
@@ -48,32 +54,175 @@ CREATE TABLE conditions (
...
+- `INSERT`: the access node sets up a multi-row prepared statement on
+  each data node and then splits the original insert statement across
+  these sub-statements. Each prepared statement allows buffering of up
+  to `timescaledb.max_insert_batch_size` number of rows (default 1000)
+  before the statement is complete and thus written to the data
+  node. Throughput can be tuned by setting a different batch size,
+  although the max size is limited by the max number of parameters of
+  a prepared statement (32767), and the number of columns in each row.
+  For example, if a distributed hypertable has 10 columns, the max
+  insert batch size is capped at 3276 rows.
```
Comment (solugebefola) [inline using-timescaledb/distributed-hypertables.md:118]: is this the max insert batch size _per data node_ or overall max?

File: using-timescaledb/distributed-hypertables.md [hunk 5]
```diff
@@ -48,32 +54,175 @@ CREATE TABLE conditions (
...
-can find more information for how to work with data in hypertables in the
-section [Creating Hypertables][creating-hypertables].
+The query performance of a distribute hypertable depends heavily
+on the ability of the planner to _push down_ computations to data
+nodes. Without push down, the query will execute on the access node by
+fetching the raw data from each data node and then performing any
+necessary computations locally. Therefore, queries that do not lead to
+push down and do not involve a lot of data nodes are unlikely to
+perform better on a distributed hypertable compared to a regular
+hypertable.
```
Comment (solugebefola) [inline using-timescaledb/distributed-hypertables.md:135]: Switch to enhancement commentary vs. limitation?
> Therefore, queries that can take advantage of push down and involve many data nodes are more likely to see better performance on a distributed hypertable compared to a regular hypertable.

File: using-timescaledb/distributed-hypertables.md [hunk 6]
```diff
@@ -48,32 +54,175 @@ CREATE TABLE conditions (
...
+  two chunks that are stored on different data nodes. The planner must
+  therefore avoid *full* push down whenever the query covers the
+  repartitioning boundary. A *partial* push down is still possible,
+  however.
+- If the query includes non-immutable functions and expressions, the
+  planner cannot push down those parts of the query since they cannot
+  be guaranteed to generate a consistent result across each data
+  node. For instance, the [`random()`][random-func] function is
+  [`VOLATILE`][volatility] since it depends on the current seed and
+  the state of the pseudo-random sequence. If the function would be
```
Comment (solugebefola) [inline using-timescaledb/distributed-hypertables.md:188]:
```suggestion
  the state of the pseudo-random sequence. If the function were to be
```

File: using-timescaledb/distributed-hypertables.md [hunk 7]
```diff
@@ -83,32 +232,88 @@ SELECT detach_data_node('node1', hypertable => 'conditions');
...
+a data node holds data that is not part of a distributed hypertable,
+this data won't be replicated. In such situations, streaming
+replication is recommended.
 
-This will prevent data from being added or updated on the data node. Once
-the data node's data is outside the retention window,
-[`drop_chunks`][drop_chunks] can be used to delete the data on the data
-node.
+To configure native replication for a distributed hypertable, set the
+parameter [`replication_factor`][replication_factor] to a value
```
Comment (solugebefola) [inline using-timescaledb/distributed-hypertables.md:294]: Not sure there is a link for this one at the footer?

File: using-timescaledb/distributed-hypertables.md [hunk 8]
```diff
@@ -48,32 +54,175 @@ CREATE TABLE conditions (
...
+
+#### Best Practices for Partitioning
+
+Like a regular hypertable, a distributed hypertable needs to be
+partitioned along a "time" dimension (e.g., a `timestamptz`
+column). However, for most distributed workloads, we recommend
+multi-dimensional partitioning with an additional "space" dimension
+that consistently partitions the data over the data nodes, similar to
+traditional
+[sharding](https://en.wikipedia.org/wiki/Shard_(database_architecture)). If
```
Comment (solugebefola) [inline using-timescaledb/distributed-hypertables.md:80]: Make into a footer link?

File: using-timescaledb/distributed-hypertables.md [hunk 9]
```diff
@@ -48,32 +54,175 @@ CREATE TABLE conditions (
...
+   GROUP BY location; 
+   ``` 
+   
+   Given that we use `location` as a space partition, each data node
+   can compute the max on its own distinct subset of the data.
+
+
+>:TIP: For push down to work, the settings variable
+>[`enable_partitionwise_aggregate`][partitionwise_agg_guc] must be set
+>to `TRUE`.
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply using-timescaledb/distributed-hypertables.md:167]: Unfortunately, it is FALSE by default. I could still add a note about that.

File: using-timescaledb/distributed-hypertables.md (hunk 6 above)
Comment (erimatnor) [reply using-timescaledb/distributed-hypertables.md:188]: shouldn't it be "If the function _was_ to be.."?

File: using-timescaledb/distributed-hypertables.md (hunk 4 above)
Comment (erimatnor) [reply using-timescaledb/distributed-hypertables.md:126]: Per data node. I'll clarify.

File: using-timescaledb/distributed-hypertables.md [hunk 10]
```diff
@@ -83,32 +232,88 @@ SELECT detach_data_node('node1', hypertable => 'conditions');
...
+fully replicating each data node using streaming replication. With
+native replication, the access node would continue to rely on
+streaming replication, however.
+
+>:WARNING: Native replication is currently under development and lacks
+>functionality for a complete high-availability solution. Some
+>functionality described in this section might not yet be implemented
+>or only have a partial feature set. For instance, there is currently
+>no mechanism implemented to re-replicate a chunk when a data node is
+>deleted due to a failure.
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply using-timescaledb/distributed-hypertables.md:264]: Added this to a separate note below.

File: using-timescaledb/distributed-hypertables.md [hunk 11]
```diff
@@ -48,32 +54,203 @@ CREATE TABLE conditions (
...
+
+## Inserting into a Distributed Hypertable [](insert)
+
+Just like a regular hypertable, it is important to batch data when
+inserting into a distributed hypertable. Each insert statement is
+often its own transaction, and the overhead and cost of this
+transaction is good to amortize over many rows of data. With a
+distributed hypertable, the transaction has additional costs due to
+the coordination that needs to happen across data nodes (e.g.,
+two-phase commit protocol). In such cases, a single INSERT transaction
```
Comment (mkindahl) [inline using-timescaledb/distributed-hypertables.md:106]: Not sure if this is intended to refer to the statement or the concept. IMHO, it's one of the two below:
```suggestion
two-phase commit protocol). In such cases, a single `INSERT` transaction
```

or 

```suggestion
two-phase commit protocol). In such cases, a single insert transaction
```

File: using-timescaledb/distributed-hypertables.md [hunk 12]
```diff
@@ -48,32 +54,203 @@ CREATE TABLE conditions (
...
+to the access node involving many rows of data is processed by the
+access node, such that the access node (a) splits the input set into
+several smaller batches of rows (with each batch having those rows
+that belong to a specific data node based on the distributed
+hypertable's partitioning), and then (b) writes each batch of rows to
+its corresponding data node.
```
Comment (mkindahl) [inline using-timescaledb/distributed-hypertables.md:112]: This is a very long sentence. It will probably be easier to read if you split it up into several.

File: using-timescaledb/distributed-hypertables.md [hunk 13]
```diff
@@ -48,32 +54,203 @@ CREATE TABLE conditions (
...
+
+There are two ways to insert data to the access node (which similarly
+uses corresponding methods when interacting with its data nodes):
```
Comment (mkindahl) [inline using-timescaledb/distributed-hypertables.md:115]:
```suggestion
There are two ways to insert data to the access node:
```

File: using-timescaledb/distributed-hypertables.md [hunk 14]
```diff
@@ -48,32 +54,203 @@ CREATE TABLE conditions (
...
+
+- `INSERT`: the access node sets up a multi-row prepared statement on
+  each data node and then splits the original insert statement across
+  these sub-statements. The access node can buffer up to
+  `timescaledb.max_insert_batch_size` number of rows (default 1000)
+  per data node before a prepared statement's limit is reached and
+  gets flushed to the data node. Thus, if there are 10000 rows in the
+  original INSERT statement and three data nodes with the default
```
Comment (mkindahl) [inline using-timescaledb/distributed-hypertables.md:123]:
```suggestion
  original `INSERT` statement and three data nodes with the default
```

File: using-timescaledb/distributed-hypertables.md [hunk 15]
```diff
@@ -48,32 +54,203 @@ CREATE TABLE conditions (
...
+  insert batch size, the insert would roughly require three full
+  batches per data node and a partial final batch.
+
+  By tuning the insert batch size, throughput can be optimized. The
+  maximum insert batch size is, however, limited by the maximum number
+  of parameters allowed in a prepared statement (32767), and the
+  number of columns in each row. For example, if a distributed
+  hypertable has 10 columns, the max insert batch size is capped at
+  3276 rows.
+- `COPY`: the access node switches each data node to COPY mode and
```
Comment (mkindahl) [inline using-timescaledb/distributed-hypertables.md:133]:
```suggestion
- `COPY`: the access node switches each data node to `COPY` mode and
```

File: using-timescaledb/distributed-hypertables.md [hunk 16]
```diff
@@ -48,32 +54,203 @@ CREATE TABLE conditions (
...
+  then routes each row to the correct data node in a stream.
+
+Out of these two options, `COPY` will typically deliver the best
+performance, although it doesn't support features like conflict
+handling (`ON CONFLICT` clause) that are used for [upserts][upserts].
+
+## Querying a Distributed Hypertable [](select)
+
+The query performance of a distribute hypertable depends heavily on
+the ability of the planner to _push down_ computations to data
```
Comment (mkindahl) [inline using-timescaledb/distributed-hypertables.md:143]: Underscores are usually not used in printed text. Emphasis or boldface is used.
```suggestion
the ability of the planner to *push down* computations to data
```

File: using-timescaledb/distributed-hypertables.md [hunk 17]
```diff
@@ -48,32 +54,203 @@ CREATE TABLE conditions (
...
+nodes. Without push down, the query will execute on the access node by
+fetching the raw data from each data node and then performing any
+necessary computations locally. Therefore, queries that can take
+advantage of push down and involve many data nodes are more likely to
+see better performance on a distributed hypertable compared to a
+regular hypertable. In particular, an aggregate computation that down
+samples data is a good candidate for push down since it both
```
Comment (mkindahl) [inline using-timescaledb/distributed-hypertables.md:150]:
```suggestion
regular hypertable. In particular, an aggregate computation that down-samples data is a good candidate for push down since it both
```

File: using-timescaledb/distributed-hypertables.md [hunk 18]
```diff
@@ -48,32 +54,203 @@ CREATE TABLE conditions (
...
+distributes the computational workload across data nodes and reduces
+the amount of data that needs to be transferred in the result set.
+
+The planner can consider two types of push downs for aggregates:
```
Comment (mkindahl) [inline using-timescaledb/distributed-hypertables.md:154]:
```suggestion
The planner can consider two types of push-downs for aggregates:
```

File: using-timescaledb/distributed-hypertables.md [hunk 19]
```diff
@@ -48,32 +54,203 @@ CREATE TABLE conditions (
...
+
+1. **FULL**: a full push down offloads all the computation to data
```
Comment (mkindahl) [inline using-timescaledb/distributed-hypertables.md:156]:
```suggestion
1. **FULL**: a full push-down offloads all the computation to data
```

File: using-timescaledb/distributed-hypertables.md [hunk 20]
```diff
@@ -48,32 +54,203 @@ CREATE TABLE conditions (
...
+   nodes and the access node only appends the individual results. To
+   enable full push down, the `GROUP BY` clause must include *all* the
+   partition keys (dimension columns) or *only* the first space
+   partition key. An example is computing the `max` temperature for
+   each location (room1, room2, etc.):
```
Comment (mkindahl) [inline using-timescaledb/distributed-hypertables.md:161]:
```suggestion
   each location (`room1`, `room2`, etc.):
```

File: using-timescaledb/distributed-hypertables.md [hunk 21]
```diff
@@ -48,32 +54,203 @@ CREATE TABLE conditions (
...
+
+   ```
+   SELECT location, max(temperature) FROM conditions
+   GROUP BY location;
+   ```
```
Comment (mkindahl) [inline using-timescaledb/distributed-hypertables.md:166]:
```suggestion
   ```sql
   SELECT location, max(temperature) FROM conditions
   GROUP BY location;
   ```
```

File: using-timescaledb/distributed-hypertables.md [hunk 22]
```diff
@@ -48,32 +54,203 @@ CREATE TABLE conditions (
...
+  pushed down to each data node, it would not generate a valid
+  pseudo-random sequence from the point of view of the access node
+  that executes the query. Another example is the
+  [`now()`][current_time] function to get the current transaction
+  time. This is a [`STABLE`][volatility] function that depends on the
+  current time zone setting on each node. TimescaleDB specifically
+  optimizes the use of `now()`, however, by constifying the function
+  on the access node, thus it is able to push down the resulting
+  timestamp to the data nodes, but this is not the case for many
+  `STABLE` functions.
```
Comment (mkindahl) [inline using-timescaledb/distributed-hypertables.md:225]: This seems to mixing several different concepts. Would be good to split it up into separate points, for example, one point about `now()` being optimized, one about handling of `VOLATILE` functions, and one about handling of `STABLE` functions.

File: using-timescaledb/distributed-hypertables.md [hunk 23]
```diff
@@ -48,32 +54,203 @@ CREATE TABLE conditions (
...
 ```
 
-Once the data node has been attached, any new chunks will also be
-distributed to the hypertable on the new data node.
+When attaching a data node, the partitioning configuration of the
+distributed hypertable is also updated to account for the additional
+data node (i.e., the number of space partitions is automatically increased to
+match), unless the function parameter `repartition` is set to
+`FALSE`. The updated partitioning configuration ensures that the data
+node will be able to take on newly created chunks.
```
Comment (mkindahl) [inline using-timescaledb/distributed-hypertables.md:253]: It is not very clear that the new data node will only be used with new chunks, that is, it will not re-partition existing chunks.

File: using-timescaledb/distributed-hypertables.md [hunk 24]
```diff
@@ -83,32 +260,143 @@ SELECT detach_data_node('node1', hypertable => 'conditions');
...
 
 Note that you cannot detach a data node that still holds data for the
-hypertable. However, if the distributed hypertable is subject to a
-[data retention policy][data-retention], it is possible to block new data
-on a data node and wait until the existing data on the data node moves
-outside the retention window. This is achieved by first
-[blocking new chunks][block_new_chunks] on the data node:
+hypertable. To be able to detach a data node, ensure that either (1)
+all its data is deleted first, or (2) the data is replicated on other
+data nodes (see the next section on native replication).
```
Comment (mkindahl) [inline using-timescaledb/distributed-hypertables.md:265]:
```suggestion
hypertable. To be able to detach a data node, ensure that either
all its data is deleted first or the data is replicated on other
data nodes (see the next section on native replication).
```

File: using-timescaledb/distributed-hypertables.md [hunk 25]
```diff
@@ -83,32 +260,143 @@ SELECT detach_data_node('node1', hypertable => 'conditions');
...
+
+The number of data nodes a chunk is written to is determined by a
+distributed hypertable's `replication_factor` (the default value is 1,
+i.e., no native replication), and it can be set when calling
+[`create_distributed_hypertable`][create_distributed_hypertable]:
+
+
+```
+SELECT create_distributed_hypertable('conditions', 'time', 'location', replication_factor => 3);
 ```
```
Comment (mkindahl) [inline using-timescaledb/distributed-hypertables.md:288]: Long line.
```suggestion
```sql
SELECT create_distributed_hypertable('conditions', 'time', 'location',
            replication_factor => 3);
```
```

File: using-timescaledb/distributed-hypertables.md [hunk 26]
```diff
@@ -83,32 +260,143 @@ SELECT detach_data_node('node1', hypertable => 'conditions');
...
+>currently no implementation for (re-)replicating existing chunks in
+>the background after a data node failure or increase in replication
+>factor. We therefore recommend keeping the replication factor set at
+>the default value of 1, and instead use streaming replication on each
+>data node.
+
+Once enabled, native replication happens as part of normal inserts, by
+writing each row to multiple data nodes, and therefore requires no
+additional mechanism for replicating newly inserted data. Existing
+chunks, however, which are not replicated according to the configured
```
Comment (mkindahl) [inline using-timescaledb/distributed-hypertables.md:306]:
```suggestion
additional mechanism for replicating newly inserted data. Existing
chunks, which are not replicated according to the configured
```

**Review by mkindahl — COMMENTED** [review]
I have added a few comments, but in general I think you should consider:

- Sentences are very long. It would benefit from splitting into separate sentences.
- You interject comments quite often. They are usually not needed and it makes the text hard to read.

File: using-timescaledb/distributed-hypertables.md [hunk 27]
```diff
@@ -48,32 +54,175 @@ CREATE TABLE conditions (
...
+that consistently partitions the data over the data nodes, similar to
+traditional
+[sharding](https://en.wikipedia.org/wiki/Shard_(database_architecture)). If
+your data set has a column called something similar to `customerID`,
+`deviceID`, or `location` (as in the example above), and it figures
+frequently in the `GROUP BY` clause of queries, then it is likely a
+good candidate column for space partitioning.
+
+With space partitioning, multi-row inserts can often be split across
+the data nodes for increased insert performance, and queries can be
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply using-timescaledb/distributed-hypertables.md:86]: There's _always_ a chance that multi-row inserts can be split across data nodes; i.e., two rows could always go to different chunks on two different data nodes, even without space partitioning.

What I want to say here is that space partitioning makes it much more likely that you will hit multiple data nodes, in particular if you are doing in-order inserts. 

I will try to clarify.

File: using-timescaledb/distributed-hypertables.md (hunk 22 above)
Comment (erimatnor) [reply using-timescaledb/distributed-hypertables.md:225]: Which concepts are you thinking of? To me this is all about the ability to push down based on the volatility of a function.

File: using-timescaledb/distributed-hypertables.md [hunk 28]
```diff
@@ -1,41 +1,47 @@
...
+nodes, the query performance of a distributed hypertable will likely
+be worse than that of a regular hypertable due to the additional
+network overhead.
+
+Note, also, that distributed hypertables can live alongside
+non-distributed tables and other objects; in fact, no objects are
+distributed by default. Interactions between distributed hypertables
+and non-distributed objects might not have the expected behavior. For
+instance, setting permissions on a distributed hypertable only works
+if the roles involved exist identically on all data nodes. Further,
```
[the remark this answers is not in the document]
Comment (erimatnor) [reply using-timescaledb/distributed-hypertables.md:29]: I wanted to put a reference, but I don't think we have a section for role management yet.

File: using-timescaledb/distributed-hypertables.md (hunk 22 above)
Comment (mkindahl) [reply using-timescaledb/distributed-hypertables.md:225]: The bullet point is talking about volatile function, stable functions, and now in different ways. It's quite hard to follow and understand what is relevant. I would suggest to break out the point about "now" being optimized and put it separately. I would also suggest to talk about volatile and stable functions separately.

File: using-timescaledb/distributed-hypertables.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (mfreed) [reply using-timescaledb/distributed-hypertables.md]: Sounds good.  A discussion later why FALSE is default =)

**Review by mfreed — APPROVED** [review]

File: using-timescaledb/distributed-hypertables.md (hunk 12 above)
Comment (erimatnor) [reply using-timescaledb/distributed-hypertables.md:112]: This sentence was proposed by Mike so I will keep for now. I think it naturally breaks up based on (a) and (b) without requiring full stops.

File: using-timescaledb/distributed-hypertables.md (hunk 13 above)
Comment (erimatnor) [reply using-timescaledb/distributed-hypertables.md:115]: Parenthesis was added based on feedback from Mike, so not reverting now.

File: using-timescaledb/distributed-hypertables.md (hunk 16 above)
Comment (erimatnor) [reply using-timescaledb/distributed-hypertables.md:143]: In markdown, this is one way to do emphasis.

File: using-timescaledb/distributed-hypertables.md (hunk 18 above)
Comment (erimatnor) [reply using-timescaledb/distributed-hypertables.md:154]: I don't think this is correct, but would defer to, e.g., Mike or Solar.

File: using-timescaledb/distributed-hypertables.md (hunk 22 above)
Comment (erimatnor) [reply using-timescaledb/distributed-hypertables.md:225]: volitile and stable functions are both non-immutable, so it makes sense to talk about them together. The important thing here is that they are _not_ immutable, so in that case I think it makes more sense to just say they're both non-immutable.

File: using-timescaledb/distributed-hypertables.md (hunk 23 above)
Comment (erimatnor) [reply using-timescaledb/distributed-hypertables.md:280]: I am intentionally not mentioning it here since it is only a current limitation. We're mentioning elsewhere that this limitation exists.

File: using-timescaledb/distributed-hypertables.md (hunk 22 above)
Comment (erimatnor) [reply using-timescaledb/distributed-hypertables.md]: I removed mentioning stable and volatile and moved the optimization for now out of the bullet list.

File: using-timescaledb/distributed-hypertables.md [hunk 29]
```diff
@@ -1,41 +1,47 @@
...
+
+>:WARNING: Distributed hypertables are not yet recommended for
+production use. For more information, please review the current
+[limitations][distributed-hypertable-limitations]. You can also
+[contact us][contact] or join the #multinode channel in our [community
+Slack][slack].
+
+## Creating a Distributed Hypertable [](create)
+
+Creating a distribute hypertable is very similar to creating a regular
```
Comment (solugebefola) [inline using-timescaledb/distributed-hypertables.md:41]:
```suggestion
Creating a distributed hypertable is very similar to creating a regular
```

File: using-timescaledb/distributed-hypertables.md (hunk 18 above)
Comment (solugebefola) [reply using-timescaledb/distributed-hypertables.md:180]: 🤷 no idea.  I think the issue is that "push down" is a verb that is contorting to be a noun here?  Maybe:
>...can consider two methods to push down aggregates

?
I'm strangely ambivalent though

File: using-timescaledb/distributed-hypertables.md [hunk 30]
```diff
@@ -48,32 +54,243 @@ CREATE TABLE conditions (
...
+regular hypertable. In particular, an aggregate computation that
+down-samples data is a good candidate for push down since it both
+distributes the computational workload across data nodes and reduces
+the amount of data that needs to be transferred in the result set.
+
+The planner can consider two types of push downs for aggregates:
+
+1. **FULL**: a full push down offloads all the computation to data
+   nodes and the access node only appends the individual results. To
+   enable full push down, the `GROUP BY` clause must include *all* the
```
Comment (solugebefola) [inline using-timescaledb/distributed-hypertables.md:184]:
```suggestion
   enable a full push down, the `GROUP BY` clause must include *all* the
```

File: using-timescaledb/distributed-hypertables.md [hunk 31]
```diff
@@ -48,32 +54,243 @@ CREATE TABLE conditions (
...
-By default, distributed hypertables will use all available data nodes when
-created. But, if you add new data nodes to the distributed database, they are
-not automatically added to existing distributed hypertables, and you need to
-explicitly attach them using [`attach_data_node`][attach_data_node].
+>:TIP: For push down to work, the settings variable
+>[`enable_partitionwise_aggregate`][partitionwise_agg_guc] must be set
+>to `TRUE` on the access node (the default is `FALSE`).
+
+Other types of push downs can include sorting operations, groupings,
+and joins. Push down for joins is currently not supported, however. To
```
Comment (solugebefola) [inline using-timescaledb/distributed-hypertables.md:215]: Yah, here too, not sure if it's a verb or noun...it seems to be used interchangeably

**Review by solugebefola — APPROVED** [review]
small things.  LGTM
